features/continue_as_new/updates_do_not_block_continue_as_new/feature.py

```diff
@@ -0,0 +1,103 @@
+module namespace, we use UnsandboxedWorkflowRunner. But this means that the workflow and
+client code execute in the same thread. Therefore we create a new thread for the client
+code to execute in, so that the two can use thread-blocking waits on the shared
+threading.Event object.
+"""
+
+import asyncio
+import threading
+from datetime import timedelta
+from uuid import uuid4
+
+from temporalio import workflow
+from temporalio.client import WorkflowHandle
+from temporalio.worker import UnsandboxedWorkflowRunner, WorkerConfig
+
+from harness.python.feature import Runner, register_feature
+from harness.python.util import admitted_update_task
+
+# See docstring at top of file.
+first_run_wft_is_in_progress = threading.Event()
+update_has_been_admitted = threading.Event()
+
+
+@workflow.defn
+class Workflow:
+    def __init__(self):
+        self.received_update = False
+
+    @workflow.run
+    async def run(self) -> str:
+        """
+        Continue as new once, then return the current run ID.
+        """
+        if not first_run_wft_is_in_progress.is_set():
+            # Note: you should usually never block the thread in workflow code.
+            # See docstring at top of file.
+            first_run_wft_is_in_progress.set()
+            update_has_been_admitted.wait()
+
+        info = workflow.info()
+        if info.continued_run_id is not None:
+            # The update is probably delivered in the first post-CAN WFT, in which case
+            # the following wait_condition is not needed. However, correct behavior does
+            # not require this to be true.
+            await workflow.wait_condition(lambda: self.received_update)
+            return info.run_id
+
+        workflow.continue_as_new()
+
+    @workflow.update
+    async def update(self) -> str:
+        """Update handler that returns the current run ID"""
+        self.received_update = True
+        return workflow.info().run_id
+
+
+async def start(runner: Runner) -> WorkflowHandle:
+    return await runner.client.start_workflow(
+        Workflow.run,
+        id=str(uuid4()),
+        task_queue=runner.task_queue,
+        execution_timeout=timedelta(minutes=1),
+    )
+
+
+async def check_result(runner: Runner, handle: WorkflowHandle) -> None:
+    async def _check_result() -> None:
+        # See docstring at top of file.
+        # Cause an update to be admitted while the first WFT is in progress
+        first_run_wft_is_in_progress.wait()
+        # The workflow is now blocking its thread waiting for the update to be admitted
+        update_task = await admitted_update_task(
+            runner.client, handle, Workflow.update, "update-id"
+        )
+        # Unblock the workflow so that it responds to the WFT with a CAN command.
+        update_has_been_admitted.set()
+        # The workflow will now CAN. Wait for the update result
+        update_run_id = await update_task
+        # The update should have been handled on the post-CAN run.
```

Comment:
I think it'd be worth two extra assertions: 1) that the update event is not in the history of the pre-CAN run, and 2) that the update is in the history of the post-CAN run

Reply:
Good call, added these.